psm/build.rs

```diff
@@ -37,24 +61,58 @@ fn main() {
     let env = ::std::env::var("CARGO_CFG_TARGET_ENV").unwrap();
     let os = ::std::env::var("CARGO_CFG_TARGET_OS").unwrap();
     let endian = ::std::env::var("CARGO_CFG_TARGET_ENDIAN").unwrap();
-    let asm = if let Some((asm, canswitch)) = find_assembly(&arch, &endian, &os, &env) {
-        println!("cargo:rustc-cfg=asm");
-        if canswitch {
-            println!("cargo:rustc-cfg=switchable_stack")
-        }
-        asm
-    } else {
-        println!(
-            "cargo:warning=Target {}-{}-{} has no assembly files!",
-            arch, os, env
-        );
-        return;
-    };
+    // Handle cross compilation scenarios where we're using eg clang-cl
+    // from a non-windows host, as by default cc will automatically try and
+    // run the appropriate Microsoft assembler for the target architecture
+    // if we give it a .asm file
+    let is_windows_host = std::env::var("HOST")
+        .unwrap_or_default()
+        .contains("-windows-");
+
+    let asm =
+        if let Some((asm, canswitch)) = find_assembly(&arch, &endian, &os, &env, is_windows_host) {
+            println!("cargo:rustc-cfg=asm");
+            if canswitch {
+                println!("cargo:rustc-cfg=switchable_stack")
+            }
+            asm
+        } else {
+            println!(
+                "cargo:warning=Target {}-{}-{} has no assembly files!",
+                arch, os, env
+            );
+            return;
+        };
 
     let mut cfg = cc::Build::new();
     let msvc = cfg.get_compiler().is_like_msvc();
 
-    if !msvc {
+    if !msvc || !is_windows_host {
+        // There seems to be a bug with clang-cl where using
+        // `/clang:-xassembler-with-cpp` is apparently accepted (ie no warnings
+        // about unused/unknown arguments), but results in the same exact error
+        // as if the flag was not present, so instead we take advantage of the
+        // fact that we're not actually compiling any C/C++ code and so don't
+        // actually need to respect any CC/CXX environment variables set to,
+        // for example, add Windows SDK include directories or the like, and if
+        // clang-cl is present, then so is clang, and we can just use it directly
+        // instead
+        if msvc && cfg.get_compiler().path().ends_with("clang-cl") {
+            // This is really dirty and I promise I feel bad about it, but
+            // cc doesn't expose a convenient way to ignore flags added from the
+            // environment :(
+            for key in
+                std::env::vars()
+                    .filter_map(|(k, _)| if k.contains("CFLAGS") { Some(k) } else { None })
+            {
+                std::env::remove_var(key);
+            }
+
+            // We need to recreate cc::Build since it caches the environment
+            cfg = cc::Build::new();
```

Comment:
I think it'd be fine to just clear the `CFLAGS` environment unconditionally.

Reply:
Yah that's a good point.